riscv/processor.cc

```diff
@@ -393,6 +393,44 @@ void state_t::reset(processor_t* const proc, reg_t max_isa)
   const reg_t henvcfg_init = (proc->extension_enabled(EXT_SVPBMT) ? HENVCFG_PBMTE : 0);
   csrmap[CSR_HENVCFG] = henvcfg = std::make_shared<henvcfg_csr_t>(proc, CSR_HENVCFG, henvcfg_mask, henvcfg_init, menvcfg);
 
+  const reg_t mstateen0_mask = (proc->extension_enabled(EXT_ZFINX) ? MSTATEEN0_FCSR : 0) |
+                               MSTATEEN0_CS | MSTATEEN0_HENVCFG | MSTATEEN_HSTATEEN;
+  csrmap[CSR_MSTATEEN0] = mstateen[0] = std::make_shared<stateen_csr_t>(proc, CSR_MSTATEEN0, mstateen0_mask, 0);
+  if (xlen == 32) {
+    const reg_t mstateen0h_mask = MSTATEEN0H_HENVCFG | MSTATEENH_HSTATEEN;
+    csrmap[CSR_MSTATEEN0H] = std::make_shared<stateenh_csr_t>(proc, CSR_MSTATEEN0H, mstateen0h_mask, mstateen[0]);
+  }
+
+  const reg_t hstateen0_mask = (proc->extension_enabled(EXT_ZFINX) ? HSTATEEN0_FCSR : 0) |
+                               HSTATEEN0_CS | HSTATEEN0_SENVCFG | HSTATEEN_SSTATEEN;
+  csrmap[CSR_HSTATEEN0] = hstateen[0] = std::make_shared<hstateen_csr_t>(proc, CSR_MSTATEEN0, hstateen0_mask, 0, 0);
+  if (xlen == 32) {
+    const reg_t hstateen0h_mask = HSTATEEN0H_SENVCFG | HSTATEENH_SSTATEEN;
+    csrmap[CSR_MSTATEEN0H] = std::make_shared<hstateenh_csr_t>(proc, CSR_HSTATEEN0H, hstateen0h_mask, hstateen[0], 0);
+  }
+
+  const reg_t sstateen0_mask = (proc->extension_enabled(EXT_ZFINX) ? SSTATEEN0_FCSR : 0) |
+                               SSTATEEN0_CS;
+  csrmap[CSR_SSTATEEN0] = sstateen[0] = std::make_shared<sstateen_csr_t>(proc, CSR_SSTATEEN0, sstateen0_mask, 0, 0);
+
+  for (int i = 1; i < 4; i++) {
```

Comment:
Once the masks are moved to the constructors, and calculated using the index, it seems this loop could cover 0-3 instead of just 1-3?

Reply:
stateen 1~3 have different mask from 0.